### PR TITLE
[MSPAINT] Introduce Ctrl+Plus / Ctrl+Minus

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing missing feature...
JIRA issue: [CORE-19094](https://jira.reactos.org/browse/CORE-19094)

## Proposed changes

- Add `Ctrl+Plus` and `Ctrl+Minus` accelerators.
- Add `IDM_CTRL_PLUS` and `IDM_CTRL_MINUS` commands.
- Stretch the selection area on `IDM_CTRL_PLUS` and `IDM_CTRL_MINUS` actions.
- Renaming: `s/SelectionModel::ClearMask/SelectionModel::ClearMaskImage/`
- Renaming: `s/SelectionModel::ClearColor/SelectionModel::ClearColorImage/`

## TODO

- [x] Do tests.

## Screenshot

image

### CODE REVIEW

File: base/applications/mspaint/lang/bg-BG.rc [hunk 1]
```diff
@@ -140,6 +140,10 @@ BEGIN
     "^I", IDM_IMAGEINVERTCOLORS
     "^E", IDM_IMAGEATTRIBUTES
     "N", IDM_IMAGEDELETEIMAGE, CONTROL, SHIFT, VIRTKEY
+    VK_ADD, IDM_CTRL_PLUS, CONTROL, VIRTKEY
+    VK_SUBTRACT, IDM_CTRL_MINUS, CONTROL, VIRTKEY
+    VK_OEM_PLUS, IDM_CTRL_PLUS, CONTROL, VIRTKEY
+    VK_OEM_MINUS, IDM_CTRL_MINUS, CONTROL, VIRTKEY
```
Comment (HBelusca) [inline base/applications/mspaint/lang/bg-BG.rc:146]: General question: I wonder whether, when these accelerators are all the same for all languages, whether they could be placed in the language-neutral section in the main mspaint .rc file :thinking:

File: base/applications/mspaint/selectionmodel.cpp [hunk 2]
```diff
@@ -540,3 +540,25 @@ void SelectionModel::UnlockBitmap(HBITMAP hbmLocked)
 {
     m_hbmColor = hbmLocked;
 }
+
+void SelectionModel::ControlPlusMinus(BOOL bMinus)
```
Comment (HBelusca) [inline base/applications/mspaint/selectionmodel.cpp:544]: A more descriptive name would be better: something like `ZoomSelection(BOOL bZoomIn)` or similar.

File: base/applications/mspaint/selectionmodel.cpp [hunk 3]
```diff
@@ -540,3 +540,25 @@ void SelectionModel::UnlockBitmap(HBITMAP hbmLocked)
...
+{
+    if (!m_bShow)
+        return;
+
+    TakeOff();
```
Comment (HBelusca) [inline base/applications/mspaint/selectionmodel.cpp:549]: What does this function do?

File: base/applications/mspaint/lang/bg-BG.rc (hunk 1 above)
Comment (katahiromz) [reply base/applications/mspaint/lang/bg-BG.rc:146]: The neutral accelerators are possible.
But some languages might want to get different key binding (?unconfirmed).

File: base/applications/mspaint/selectionmodel.cpp (hunk 2 above)
Comment (katahiromz) [reply base/applications/mspaint/selectionmodel.cpp:544]: This feature actually stretches the selection.
Zooming is visual only.

File: base/applications/mspaint/selectionmodel.cpp (hunk 3 above)
Comment (katahiromz) [reply base/applications/mspaint/selectionmodel.cpp:549]: The selected area will take off.
(The image of selected area is transported)

File: base/applications/mspaint/lang/bg-BG.rc (hunk 1 above)
Comment (katahiromz) [reply base/applications/mspaint/lang/bg-BG.rc:146]: If there are language-specific accelerators, they will override the neutral.

File: base/applications/mspaint/selectionmodel.cpp (hunk 2 above)
Comment (HBelusca) [reply base/applications/mspaint/selectionmodel.cpp:544]: Ok then maybe call the function "StretchSelection"

Comment (katahiromz) [reply base/applications/mspaint/selectionmodel.cpp]: Done.

File: base/applications/mspaint/lang/bg-BG.rc (hunk 1 above)
Comment (katahiromz) [reply base/applications/mspaint/lang/bg-BG.rc:146]: For example, `pt-PT.rc` has line `"^O", IDM_FILENEW`.

File: base/applications/mspaint/selectionmodel.cpp (hunk 3 above)
Comment (katahiromz) [reply base/applications/mspaint/selectionmodel.cpp:554]: Readability is improved.

File: base/applications/mspaint/lang/bg-BG.rc (hunk 1 above)
Comment (HBelusca) [reply base/applications/mspaint/lang/bg-BG.rc:146]: good catch :+1:

File: base/applications/mspaint/selectionmodel.cpp [hunk 4]
```diff
@@ -540,3 +545,26 @@ void SelectionModel::UnlockBitmap(HBITMAP hbmLocked)
...
+
+    INT cx = m_rc.Width(), cy = m_rc.Height();
+
+    if (bShrink)
+        m_rc.InflateRect(-cx / 4, -cy / 4);
+    else
+        m_rc.InflateRect(+cx / 2, +cy / 2);
+
+    // The selection area must exist at here
```
Comment (HBelusca) [inline base/applications/mspaint/selectionmodel.cpp:563]:
```suggestion
    // The selection area must exist there
```

File: base/applications/mspaint/winproc.cpp [hunk 5]
```diff
@@ -1069,6 +1069,13 @@ LRESULT CMainWindow::OnCommand(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL& bH
             fullscreenWindow.DoCreate();
             fullscreenWindow.ShowWindow(SW_SHOWMAXIMIZED);
             break;
+
+        case IDM_CTRL_PLUS:
+            toolsModel.ControlPlusMinus(FALSE);
+            break;
+        case IDM_CTRL_MINUS:
+            toolsModel.ControlPlusMinus(TRUE);
```
Comment (HBelusca) [inline base/applications/mspaint/winproc.cpp:1077]: Those need to name change as well.

File: base/applications/mspaint/toolsmodel.cpp [hunk 6]
```diff
@@ -248,3 +248,8 @@ void ToolsModel::selectAll()
     OnMouseMove(TRUE, imageModel.GetWidth(), imageModel.GetHeight());
     OnButtonUp(TRUE, imageModel.GetWidth(), imageModel.GetHeight());
 }
+
+void ToolsModel::ControlPlusMinus(BOOL bMinus)
+{
+    m_pToolObject->OnControlPlusMinus(bMinus);
```
Comment (HBelusca) [inline base/applications/mspaint/toolsmodel.cpp:254]: Name change needed

File: base/applications/mspaint/winproc.cpp (hunk 5 above)
Comment (katahiromz) [reply base/applications/mspaint/winproc.cpp:1077]: What name?
`SpecialAction`?

Comment (katahiromz) [reply base/applications/mspaint/winproc.cpp:1077]: `AdjustPlusMinus`

Comment (HBelusca) [reply base/applications/mspaint/winproc.cpp:1077]: Hmmmm... now that you're asking: Do you think the "CtrlPlusMinus" thing can do different things depending on which tool is currently selected?

Comment (katahiromz) [reply base/applications/mspaint/winproc.cpp:1077]: `SpecialTweak`

Comment (katahiromz) [reply base/applications/mspaint/winproc.cpp:1077]: Yes, it can do different jobs.
https://github.com/katahiromz/RPaint/issues/1

Comment (katahiromz) [reply base/applications/mspaint/winproc.cpp]: Renamed as `SpecialTweak`.

Comment (HBelusca) [reply base/applications/mspaint/winproc.cpp]: OK!

**Review by HBelusca — APPROVED** [review]

**Pull request merged**